src/FreetypeRenderer.cc

```diff
@@ -196,31 +244,46 @@ void FreetypeRenderer::detect_properties(FreetypeRenderer::Params &params) const
 	params.set_direction(hb_direction_to_string(direction));
 }
 
-std::vector<const Geometry *> FreetypeRenderer::render(const FreetypeRenderer::Params &params) const
+FT_Face FreetypeRenderer::get_font_face(const FreetypeRenderer::Params &params)
 {
-	FT_Face face;
-	FT_Error error;
-	DrawingCallback callback(params.segments, params.size);
-	
-	FontCache *cache = FontCache::instance();
-	if (!cache->is_init_ok()) {
-		return std::vector<const Geometry *>();
-	}
+    FontCache *cache = FontCache::instance();
+    if (!cache->is_init_ok()) {
+        LOG(message_group::Warning,params.loc,params.documentPath,"Font cache initialization failed");
+        return nullptr;
+    }
+
+    FT_Face face = cache->get_font(params.font);
+    if (face == nullptr) {
+        LOG(message_group::Warning,params.loc,params.documentPath,"Can't get font %1$s", params.font);
+        return nullptr;
+    }
+    
+    FT_Error error = FT_Set_Char_Size(face, 0, scale, 100, 100);
+    if (error) {
+        LOG(message_group::Warning,params.loc,params.documentPath,"Can't set font size for font %1$s",params.font);
+        return nullptr;
+    }
+    return face;
+}
 
-	face = cache->get_font(params.font);
-	if (face == nullptr) {
-		return std::vector<const Geometry *>();
-	}
-	
-	error = FT_Set_Char_Size(face, 0, scale, 100, 100);
-	if (error) {
-		LOG(message_group::None,Location::NONE,"","Can't set font size for font %1$s",params.font);
-		return std::vector<const Geometry *>();
-	}
-	
-	hb_font_t *hb_ft_font = hb_ft_font_create(face, nullptr);
+FreetypeRenderer::ShapeResults::ShapeResults()
+{
+    hb_ft_font = nullptr;
+    hb_buf = nullptr;
+}
 
-	hb_buffer_t *hb_buf = hb_buffer_create();
+bool FreetypeRenderer::ShapeResults::shape(
+    const FreetypeRenderer::Params &params)
+{
+    FT_Error error;
+    FT_Face face = get_font_face(params);
+    if (face == nullptr) {
+        return false;
+    }
+
+	hb_ft_font = hb_ft_font_create(face, nullptr);
```

Comment:
`shape()` is only ever called once.

The buffer cleanup can't be in shape(), because one of shape()'s results is `glyph_array`, and `glyph_array`'s `GlyphData` members include a pointer to the position data that comes back from `hb_buffer_get_glyph_positions()`.  The HB data has to stick around until `text()` is done using `glyph_array`.

I was unhappy about that asymmetry too.  I considered doing the real work in the constructor, but there's no good way to return an error from a constructor.

In the old implementation, all of the shape calculation is in `FreetypeRenderer::render` using local variables.  Here, they're extracted out into `ShapeResults::shape`, and are consumed by both `FreetypeRenderer::render()` and `FreetypeRenderer::TextMetrics::calculate()`.  Moving the HB pointers into ShapeResults meant that ShapeResults needed to clean them up.